### PR TITLE
Reformatted position control and added rotation control

### PR DESCRIPTION
**Pull request by OliverW10**
Changed the existing position and colour reading code to follow the new structure and added code to complete rotation control.

### CODE REVIEW

File: components/spinner.py [hunk 1]
```diff
@@ -2,30 +2,129 @@
 import rev
 import rev.color
 
+class Colour: #class for storing rgb colours
+    def __init__(self, red: float, green: float, blue: float):
+        self.red = red
+        self.green = green
+        self.blue = blue
+
+    def dist(self, c) -> tuple:
```
Comment (auscompgeek) [inline components/spinner.py]: Is this type hint correct?

File: components/spinner.py [hunk 2]
```diff
@@ -2,30 +2,129 @@
...
+        return (
+            abs(self.red - c.red) + abs(self.green - c.green) + abs(self.blue - c.blue)
+        )
 
 class Spinner:
-    spinner_motor: wpilib.Spark
-    spinner_solenoid: wpilib.DoubleSolenoid
-    colour_sensor: rev.color.ColorSensorV3
+	spinner_motor: wpilib.Spark
```
Comment (auscompgeek) [inline components/spinner.py]: Inconsistent indentation. Please ensure your indentation is consistent within the same file.

Comment (auscompgeek) [reply components/spinner.py]: (Preferably, run black.)

Comment (OliverW10) [reply components/spinner.py]: I have now run black

File: controllers/spinner.py [hunk 3]
```diff
@@ -58,53 +28,14 @@ def select_task(self):
...
-            return 0
-        distance = (
-            self.WHEEL_ORDER.index(self.required_colour)
-            + self.COLOUR_OFFSET
-            - self.WHEEL_ORDER.index(current_colour)
-        )
-        if distance > 2:
-            return distance - 4
-        return distance
+        pass
```
Comment (mlists) [inline controllers/spinner.py]: This is not a component and probably doesn't need an execute function, especially an empty one.

File: components/spinner.py [hunk 4]
```diff
@@ -3,29 +3,131 @@
...
+
+    def run_rotation(self) -> None:  # rotation control to be called every 20ms
+        self.spinner_motor.set(
+            1 - self.rotations / 6
+        )  # speed slows down as it gets closer to required revolutions
+        current_colour = self.get_current_colour()
+        if self.lastCol != current_colour:
+            self.rotations += 1 / 8
+        self.lastCol = current_colour
+        if self.rotations >= 3.5 and self.read_colour()[1] > 35:
```
Comment (mlists) [inline components/spinner.py]: Consider using the piston state or a variable to track whether the mechanism is up or down, if you keep this method add a comment to explain what you are doing and why as it is currently not obvious

File: components/spinner.py [hunk 5]
```diff
@@ -3,29 +3,131 @@
...
+        self.spinner_solenoid.set(False)
+
+    def read_colour(self) -> tuple:
         return (self.colour_sensor.getColor(), self.colour_sensor.getIR())
 
-    def piston_up(self):
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kForward)
+    def get_wheel_dist(
+        self,
+    ) -> int:  # finds the colour wheels distance from required colour in segments
```
Comment (auscompgeek) [inline components/spinner.py]: Why does this simple method signature span multiple lines?

Comment (OliverW10) [reply components/spinner.py]: ask black

File: components/spinner.py (hunk 4 above)
Comment (OliverW10) [reply components/spinner.py]: I made the suggested changes

File: controllers/spinner.py (hunk 3 above)
Comment (OliverW10) [reply controllers/spinner.py]: I deleted the execute for the controller

File: components/spinner.py [hunk 6]
```diff
@@ -3,29 +3,131 @@
...
+    def __init__(self) -> None:
+        self.rotations = 0
+        self.state = None
+        self.lastCol = None
+        self.required_colour = None
+
+    def on_enable(self) -> None:
         self.spinner_motor.stopMotor()
-        self.spinner_solenoid.set(wpilib.DoubleSolenoid.Value.kReverse)
+        self.spinner_solenoid.set(True)
```
[the remark this answers is not in the document]
Comment (OliverW10) [reply components/spinner.py]: It now does

File: components/spinner.py (hunk 5 above)
Comment (auscompgeek) [reply components/spinner.py]: Are you going to fix it?

File: components/shooter.py [hunk 7]
```diff
@@ -89,6 +89,10 @@ def is_firing(self) -> bool:
         """
         return not self.loading_piston.get()
 
+    def set_motor_rpm(self, v_outer, v_centre) -> None:
+        self.centre_rpm = v_centre
+        self.outer_rpm = v_outer
+
```
Comment (mlists) [inline components/shooter.py]: This seems to have slipped through one of your merges

**Pull request closed without merging**